src/backend/distributed/planner/multi_planner_utils.c

```diff
@@ -0,0 +1,863 @@
+
+		commonRteIdentities = bms_add_member(commonRteIdentities, rteIdentity);
+	}
+
+	/* check whether all relations exists in the main restriction list */
+	foreach(relationRestrictionCell, restrictionContext->relationRestrictionList)
+	{
+		RelationRestriction *relationRestriction =
+			(RelationRestriction *) lfirst(relationRestrictionCell);
+		int rteIdentity = GetRTEIdentity(relationRestriction->rte);
+
+		if (PartitionKey(relationRestriction->relationId) &&
+			!bms_is_member(rteIdentity, commonRteIdentities))
+		{
+			return false;
+		}
+	}
+
+	return true;
+}
+
+
+/*
+ * ReferenceRelationCount iterates over the relations and returns the reference table
+ * relation count.
+ */
+static uint32
+ReferenceRelationCount(RelationRestrictionContext *restrictionContext)
+{
+	ListCell *relationRestrictionCell = NULL;
+	uint32 referenceRelationCount = 0;
+
+	foreach(relationRestrictionCell, restrictionContext->relationRestrictionList)
+	{
+		RelationRestriction *relationRestriction =
+			(RelationRestriction *) lfirst(relationRestrictionCell);
+
+		if (PartitionMethod(relationRestriction->relationId) == DISTRIBUTE_BY_NONE)
+		{
+			referenceRelationCount++;
+		}
+	}
+
+	return referenceRelationCount;
+}
+
+
+/*
+ * GenerateAttributeEquivalencesForRelationRestrictions gets a relation restriction
+ * context and returns a list of AttributeEquivalenceClass.
+ *
+ * The algorithm followed can be summarized as below:
+ *
+ * - Per relation restriction
+ *     - Per plannerInfo's eq_class
+ *         - Create an AttributeEquivalenceClass
+ *         - Add all Vars that appear in the plannerInfo's
+ *           eq_class to the AttributeEquivalenceClass
+ *               - While doing that, consider LATERAL vars as well.
+ *                 See GetVarFromAssignedParam() for the details. Note
+ *                 that we're using parentPlannerInfo while adding the
+ *                 LATERAL vars given that we rely on that plannerInfo.
+ *
+ */
+static List *
+GenerateAttributeEquivalencesForRelationRestrictions(RelationRestrictionContext
+													 *restrictionContext)
+{
+	List *attributeEquivalenceList = NIL;
+	ListCell *relationRestrictionCell = NULL;
+
+	foreach(relationRestrictionCell, restrictionContext->relationRestrictionList)
+	{
+		RelationRestriction *relationRestriction =
+			(RelationRestriction *) lfirst(relationRestrictionCell);
+		List *equivalenceClasses = relationRestriction->plannerInfo->eq_classes;
+		ListCell *equivalenceClassCell = NULL;
+
+		foreach(equivalenceClassCell, equivalenceClasses)
+		{
+			EquivalenceClass *plannerEqClass =
+				(EquivalenceClass *) lfirst(equivalenceClassCell);
+
+			AttributeEquivalenceClass *attributeEquivalance =
+				AttributeEquivalenceClassForEquivalenceClass(plannerEqClass,
+															 relationRestriction);
+
+			attributeEquivalenceList =
+				AddAttributeClassToAttributeClassList(attributeEquivalenceList,
+													  attributeEquivalance);
+		}
```

Comment:
The filtering is happening on the `AddAttributeClassToAttributeClassList ()`, so we're not generating the same thing over and over. But, the algorithm that does the filtering is o(n^3). 

Note that this O(n^3) only includes partition keys, we're not adding all the restrictions to these lists. That's why I'm not very much concerned about that. Do you want me to work and improve that algorithm? Maybe using a hash table/bms or so?